File: Data Structures & Algorithms/depth-of-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        # Return 0 if the root is None
        if not root:
            return 0

        # Recursively compute the depth of the left/right subtrees and add 1 for each of them
        left_depth = 1 + self.maxDepth(root.left)
        right_depth = 1 + self.maxDepth(root.right)

        # Return the maximum from left/right depths as we unwind
        return max(left_depth, right_depth)

File: Data Structures & Algorithms/same-binary-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
   public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        queue<TreeNode*> queueP;
        queue<TreeNode*> queueQ;
        queueP.push(p);
        queueQ.push(q);

        while (!queueP.empty() && !queueQ.empty()) {
            TreeNode* nodeP = queueP.front();
            TreeNode* nodeQ = queueQ.front();
            queueP.pop();
            queueQ.pop();

            if (nodeP == nullptr && nodeQ == nullptr) {
                continue;
            }
            if (nodeP == nullptr || nodeQ == nullptr) {
                return false;
            }
            if (nodeP->val != nodeQ->val) {
                return false;
            }
            queueP.push(nodeP->left);
            queueP.push(nodeP->right);
            queueQ.push(nodeQ->left);
            queueQ.push(nodeQ->right);
        }
        return true;
    }
};
